src/storages/pluggable/SplitsCachePluggable.ts

```diff
@@ -0,0 +1,274 @@
+import { isFiniteNumber, toNumber, isNaNNumber } from '../../utils/lang';
+import KeyBuilder from '../KeyBuilder';
+import { ICustomStorageWrapper, ISplitsCacheAsync } from '../types';
+import { ILogger } from '../../logger/types';
+import { usesSegments } from '../AbstractSplitsCacheSync';
+import { ISplit } from '../../dtos/types';
+import { logPrefix } from './constants';
+import { SplitError } from '../../utils/lang/errors';
+
+/**
+ * ISplitsCacheAsync implementation for pluggable storages.
+ */
+export class SplitsCachePluggable implements ISplitsCacheAsync {
+
+  private readonly log: ILogger;
+  private readonly keys: KeyBuilder;
+  private readonly wrapper: ICustomStorageWrapper;
+
+  /**
+   *
+   * @param log logger instance
+   * @param keys key builder
+   * @param wrapper adapted wrapper storage
+   */
+  constructor(log: ILogger, keys: KeyBuilder, wrapper: ICustomStorageWrapper) {
+    this.log = log;
+    this.keys = keys;
+    this.wrapper = wrapper;
+  }
+
+  private _decrementCounts(split: ISplit) {
+    const promises = [];
+    if (split.trafficTypeName) {
+      const ttKey = this.keys.buildTrafficTypeKey(split.trafficTypeName);
+      promises.push(this.wrapper.decr(ttKey));
+    }
+
+    if (usesSegments(split)) {
+      const segmentsCountKey = this.keys.buildSplitsWithSegmentCountKey();
+      promises.push(this.wrapper.decr(segmentsCountKey));
+    }
+
+    return Promise.all(promises);
+  }
+
+  private _incrementCounts(split: ISplit) {
+    const promises = [];
+    if (split.trafficTypeName) {
+      const ttKey = this.keys.buildTrafficTypeKey(split.trafficTypeName);
+      promises.push(this.wrapper.incr(ttKey));
+    }
+
+    if (usesSegments(split)) {
+      const segmentsCountKey = this.keys.buildSplitsWithSegmentCountKey();
+      promises.push(this.wrapper.incr(segmentsCountKey));
+    }
+
+    return Promise.all(promises);
+  }
+
+  /**
+   * Add a given split.
+   * The returned promise is resolved when the operation success
+   * or rejected with an SplitError if it fails (e.g., wrapper operation fails)
+   */
+  addSplit(name: string, split: string): Promise<boolean> {
+    const splitKey = this.keys.buildSplitKey(name);
+    return this.wrapper.get(splitKey).then(splitFromStorage => {
+
+      // handling parsing error as SplitErrors
+      let parsedPreviousSplit, parsedSplit;
+      try {
+        parsedPreviousSplit = splitFromStorage ? JSON.parse(splitFromStorage) : undefined;
+        parsedSplit = JSON.parse(split);
+      } catch (e) {
+        throw new SplitError('Error parsing split definition: ' + e);
+      }
+
+      return Promise.all<boolean[], boolean, boolean[]>([
+        // If it's an update, we decrement the traffic type and segment count of the existing split,
+        parsedPreviousSplit && this._decrementCounts(parsedPreviousSplit),
```

Comment:
Are we handling correct the scenario in which `parsedPreviousSplit` is `undefined`?